### ISSUE TITLE
Output of `torch.sum` for Float16 deviates more than tolerance from numpy on CPU

### ISSUE THREAD

**Issue by imaginary-person** [pytorch/pytorch#59489]
## 🐛 Bug

`torch.sum` for float16 deviates (more than the tolerance) from numpy.

## To Reproduce

1. For the following example of `sum`,
PyTorch CPU implementation outputs 4444.0.
PyTorch CUDA sums it as 4448.0. (FWIW, PyTorch doesn't guarantee the same results on CPU & CUDA).
`numpy 1.20.3` sums it as 4450.0.

```
import torch
import numpy
x = torch.tensor([[66, 63, 21, 84, 86, 90, 86, 21],
                  [81, 68, 22, 39, 43, 38, 24, 63],
                  [96, 98, 33, 85, 90, 87, 52, 99],
                  [56, 21, 95, 59, 85, 51, 16, 37],
                  [74, 55, 95, 97, 45, 82, 71, 33],
                  [88, 90, 91, 68, 31, 22, 53, 30],
                  [68, 22, 73, 82, 36, 56, 96, 20],
                  [71, 66, 82, 17, 97, 35, 20, 43],
                  [31, 81, 87, 90, 60, 49, 96, 91]], dtype=torch.half)
print(torch.sum(x, dtype=torch.half))
print(numpy.nansum(x.detach().cpu().numpy(), dtype=numpy.float16))
# torch outputs 4444.0 on CPU, but numpy's output is 4450.0
# PyTorch BFloat16 has the same output on both CPU & CUDA (4448.0)

```

## Expected behavior

Can the accuracy can be improved for `torch.sum` of `torch.half` on CPU?
Currently, all floating types share the same CPU implementation of `sum` in `SumKernel.cpp`. 
Could the half-types benefit from a slightly different approach?
Thanks!

## Environment

Built from current master branch's source-code

OS: Ubuntu 20.04 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8 (64-bit runtime)
Python platform: Linux-5.4.0-67-generic-x86_64-with-glibc2.29

cc @mruberry @rgommers @heitorschueroff @peterbell10
